src/common/io.cc

```diff
@@ -6,14 +6,96 @@
 #include <fcntl.h>
 #include <unistd.h>
 #endif  // defined(__unix__)
+#include <algorithm>
 #include <cstdio>
 #include <string>
 
 #include "xgboost/logging.h"
+#include "io.h"
 
 namespace xgboost {
 namespace common {
 
+size_t PeekableInStream::Read(void* dptr, size_t size) {
+  size_t nbuffer = buffer_.length() - buffer_ptr_;
+  if (nbuffer == 0) return strm_->Read(dptr, size);
+  if (nbuffer < size) {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, nbuffer);
+    buffer_ptr_ += nbuffer;
+    return nbuffer + strm_->Read(reinterpret_cast<char*>(dptr) + nbuffer,
+                                 size - nbuffer);
+  } else {
+    std::memcpy(dptr, dmlc::BeginPtr(buffer_) + buffer_ptr_, size);
+    buffer_ptr_ += size;
+    return size;
+  }
+}
+
+size_t PeekableInStream::PeekRead(void* dptr, size_t size) {
+  size_t nbuffer = buffer_.length() - buffer_ptr_;
+  if (nbuffer < size) {
+    buffer_ = buffer_.substr(buffer_ptr_, buffer_.length());
```

Comment:
Line 37-41 seems to be just a potential optimization (shrink_to_fit).
I suggest to consider
1. extracting to a separate method so it would be easier to follow the logic
2. potentially removing it cause it may lead to unexpected to the client actions

If you want you can give more control to the client by exposing the method shrink

What do you think?

Reply:
@trams It's copied code from header to avoid "recompiling the world".  It troubles me that `dmlc-core` has a complete different framework for IO with self defined streams instead of aligning with standard c++ or c.  Will think about your suggestion.  Thanks for the suggestion!